File: rust/fib.rs
use std::env;

fn fib(n: usize) -> usize {
  if n < 2 {
    n
  } else {
    fib(n - 1) + fib(n - 2)
  }
}

fn fib_pattern(n: usize) -> usize {
  match n {
    0 => 0,
    1 => 1,
    n => fib(n - 1) + fib(n - 2),
  }
}

fn fib_tail_recursive(n: usize) -> usize {
  fn fib_tail_iter(n: usize, prev_fib: usize, fib: usize) -> usize {
    match n {
      0 => prev_fib,
      n => fib_tail_iter(n - 1, fib, prev_fib + fib),
    }
  }
  fib_tail_iter(n, 0, 1)
}

fn fib_linear(mut n: usize) -> usize {
  if n < 2 {
    n
  } else {
    let mut prev_fib = 0;
    let mut fib = 1;
    while n > 1 {
      let temp = prev_fib + fib;
      prev_fib = fib;
      fib = temp;
      n -= 1
    }
    fib
  }
}

fn main() {
  let args: Vec<_> = env::args().collect();
  let input = if args.len() > 1 {
    args[1].parse::<usize>().unwrap()
  } else {
    29
  };
  println!("{}", fib_linear(input));
}

File: perl/fib.pl
#! /usr/bin/perl
use strict;
use warnings;
use Math::BigInt;
use Math::BigFloat;

sub fib {
  my $n = shift;
  $n < 2 ? $n : fib($n - 1) + fib($n - 2);
}

sub fibLinear {
  my ($n, $prevFib, $fib) = (shift, 0, 1);
  ($prevFib, $fib) = ($fib, $prevFib + $fib) while --$n > 0;
  $fib;
}

sub fibFormula {
  my $n = Math::BigFloat->new(shift);
  $n = ((5 ** 0.5 + 1) / 2.0) ** $n / 5 ** 0.5;
  $n->bfround(1);
}

sub fibTailAuxiliary {
  my ($n, $prevFib, $fib) = (@_);
  $n == 0 ? $prevFib : fibTailAuxiliary($n - 1, $fib, $prevFib + $fib);
}
sub fibTailRecursive {
  fibTailAuxiliary(shift, 0, 1);
}

sub fibLinearBig {
  my $n = Math::BigInt->new(shift);
  my $prevFib = Math::BigInt->new(0);
  my $fib = Math::BigInt->new(1);
  ($prevFib, $fib) = ($fib, $prevFib + $fib) while --$n > 0;
  $fib;
}

my $input = shift;
unless (defined $input) {
  $input = 29;
}
printf("Fibonacci %d: %d\n", $input, fibTailRecursive($input));
